qiime2/core/cache.py

```diff
@@ -0,0 +1,162 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2022, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import os
+import yaml
+import pathlib
+
+import qiime2
+from qiime2.sdk.result import Artifact
+
+_VERSION_TEMPLATE = """\
+QIIME 2
+cache: %s
+framework: %s
+"""
+
+_KEY_TEMPLATE = """\
+origin:
+ %s
+data:
+ %s
+pool:
+ %s
+"""
+
+
+class Cache:
+    """General structure of the cache (tmp optional)
+    artifact_cache/
+    ├── data
+    │   ├── uuid1
+    │   ├── uuid2
+    │   ├── uuid3
+    │   └── uuid4
+    ├── keys
+    │   ├── bar.yaml
+    │   ├── baz.yaml
+    │   └── foo.yaml
+    ├── pools
+    │   └── puuid1
+    │       ├── uuid2 -> ../../data/uuid2/
+    │       └── uuid3 -> ../../data/uuid3/
+    ├── tmp
+    └── VERSION
+    """
+    CURRENT_FORMAT_VERSION = '1'
+
+    def __init__(self, path):
+        self.path = pathlib.Path(path)
+
+        # Do we want a more rigorous check for whether or not we've been
+        # pointed at an existing cache?
+        if not os.path.exists(self.path):
+            self.create_cache()
+        elif not self.is_cache():
+            raise ValueError(f"Path: \'{path}\' already exists and is not a"
+                             " cache")
+
+    # Surely this needs to be a thing? I suppose if they hand us a path that
+    # doesn't exist we just create a cache there? do we want to create it at
+    # that exact path do we slap a 'cache' sub-directory at that location and
+    # use that?
+    def create_cache(self):
+        os.mkdir(self.path)
+        os.mkdir(self.path / 'data')
+        os.mkdir(self.path / 'keys')
+        os.mkdir(self.path / 'pools')
+        # Do we want this right off the bat? How exactly is setting tmp in the
+        # cache going to work? tmp is never going to be managed by the cache,
+        # it's just so they're both on the same disk, so they'll probably just
+        # set the tmp location in the config or something. I feel like if we're
+        # going to manage the cache, we should manage the cache which means if
+        # they're going to put tmp in the cache it should have to be in a set
+        # directory within the cache like tmp not just whatever they want it to
+        # be in the cache. Not sure how we would really enforce that, but we
+        # can just... Heavily encourage it I guess
+        # os.mkdir('tmp')
+
+        self.version.write_text(
+            _VERSION_TEMPLATE % (self.CURRENT_FORMAT_VERSION,
+                                 qiime2.__version__))
+
+    # Tell us if the path is a cache or not
+    def is_cache(self):
```

Comment:
You'll probably want this to be a class method, which can be invoked on the Class object, or an instance.

```python
@classmethod
def is_cache(cls, path):
    pass
```

Instead of recieving `self` you will receive your class back (or the relevant subclass).

Reply:
Good idea